h/claim/__init__.py

```diff
@@ -0,0 +1,13 @@
+from webob.cookies import SignedSerializer
+
+
+def includeme(config):
+    config.include('.views')
+    config.include('.util')
+
+    # We use the shared session secret, but salt it with the namespace
+    # 'h.claim' -- only messages serialized with this salt will
+    # authenticate on deserialization.
+    secret = config.registry.settings['session.secret']
+    serializer = SignedSerializer(secret, 'h.claim')
+    config.registry.claim_serializer = serializer
```

Comment:
I'm not suggesting to remove the h.claim string, just to use a different
secret.

If the operator sets it to the same, your concatenation is the failsafe.

But better for them to set it differently.

Not sure about your derivation idea. Will think after coffee. But I'm
pretty sure just using a different key is the easiest to reason about.
On Feb 26, 2015 2:05 AM, "Nick Stenning" <notifications@github.com> wrote:

> In h/claim/__init__.py
> <https://github.com/hypothesis/h/pull/1941#discussion_r25415917>:
>
> > @@ -0,0 +1,13 @@
> > +from webob.cookies import SignedSerializer
> > +
> > +
> > +def includeme(config):
> > +    config.include('.views')
> > +    config.include('.util')
> > +
> > +    # We use the shared session secret, but salt it with the namespace
> > +    # 'h.claim' -- only messages serialized with this salt will
> > +    # authenticate on deserialization.
> > +    secret = config.registry.settings['session.secret']
> > +    serializer = SignedSerializer(secret, 'h.claim')
> > +    config.registry.claim_serializer = serializer
>
> Thanks for asking the question. It's good to be paranoid about such
> things. Just to be clear what this is doing. The "salt" isn't really a salt
> in any conventional sense. Usually a salt is concatenated to the message in
> a hashing (or possibly symmetric encryption) operation:
>
>     h(M || S)
>     E(k, M || S)
>
> In this context, the "salt" is actually concatenated to the *key*, in
> order to form a new key, such that a message M serialized with salt S_1
> will not authenticate when passed to the deserializer using salt S_1.
> That is:
>
>     HMAC(k || S_1, M) != HMAC(k || S_2, M)
>
> So, the question is: how safe is it to derive keys by concatenating a
> known string to them? In general, the answer is "not very," but for the
> specific case of HMAC it's probably ok. As far as I'm aware there aren't
> currently any key recovery or forgery attacks against HMAC-SHA1 anyway.
>
> All of that said, a general cryptographic principle says that keys should
> be uniform random, so I think I will rework this to derive the keys for
> these serializers using HMAC instead, so that instead of k || S, we use HMAC(k,
> S) as the derived keys for each purpose.
>
> Why not get the deployer to provide multiple keys? Because the faff
> associated with setting up N keys (the number will only grow over time)
> actually makes it likely that at some point someone will simply set all the
> keys to the same thing, which would be *more* dangerous than what we have
> currently. Or, to put it another way, if we ask deployers to provide *one*
> strong secret key, then we can derive keys that are *guaranteed*
> different for different purposes as:
>
>     k_session = HMAC(k, S_session)
>     k_claim = HMAC(k, S_claim)
>     ...
>
> Whereas if we ask deployers to provide N different secret keys, we will
> have no guarantee that the supplied keys are different.
>
> TL;DR: You're probably being overly paranoid, but that's good. I'll rework
> this to derive keys properly.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/hypothesis/h/pull/1941/files#r25415917>.
>